algviz/parsers/structures.py

```diff
@@ -0,0 +1,207 @@
+        self.data = [obj_table[elt] for elt in self.data]
+
+class _Singleton(abc.ABCMeta):
+    # Embarassingly copied from https://stackoverflow.com/questions/6760685
+    _instances = {}
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(_Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+class NullType(DataStructure, metaclass=_Singleton):
+    """A null value.  Similar to python's None.  Fun to make."""
+    def __init__(self):
+        super().__init__(uid="#null")
+        del self.metadata  # no static dictionary allowed
+
+    def untablify(self, obj_table):
+        pass
+
+    def __bool__(self):
+        return False
+
+    def __repr__(self):
+        return "Null"
+
+Null = NullType()
+
+class LinkedListNode(DataStructure):
+    def __init__(self, value, successor=Null):
+        self.value = value
+        self.successor = successor
+
+    def untablify(self, obj_table):
+        self.value = obj_table[self.value]
+        self.successor = obj_table[self.successor]
+
+class Graph(DataStructure):
+    def __init__(self, nodes, edges, **kwargs):
+        super().__init__(**kwargs)
+        self.nodes = nodes
+        self.edges = edges
+
+    def untablify(self, obj_table):
+        self.nodes = [obj_table[n] for n in self.nodes]
+        self.edges = [obj_table[e] for e in self.edges]
+
+class Node(DataStructure):
+    def __init__(self, data, **kwargs):
+        super().__init__(**kwargs)
+        self.data = data
+        # self.edges = edges  # no storing edges, I guess.  TODO: discuss this with people
+
+    def untablify(self, obj_table):
+        self.data = obj_table[self.data]
+        # self.edges = list(obj_table[e] for e in self.edges)
+
+class Edge(DataStructure):
+    def __init__(self, orig, dest, data=Null, **kwargs):
+        super().__init__(**kwargs)
+        self.orig = orig
+        self.dest = dest
+        self.data = data
+
+    def untablify(self, obj_table):
+        self.orig = obj_table[self.orig]
+        self.dest = obj_table[self.dest]
+        self.data = obj_table[self.data]
+
+class Widget(DataStructure):
+    # Potentially our most useful DataStructure
+
+    def __eq__(self, other):
+        return isinstance(other, Widget)
+
+    def untablify(self, obj_table):
+        pass
+
+class BinaryTree(DataStructure):
+    def __init__(self, root, **kwargs):
+        super().__init__(**kwargs)
+        self.root = root
+
+    def untablify(self, obj_table):
+        self.root = obj_table[self.root]
+
+    def __eq__(self, other):
+        return isinstance(other, BinaryTree) and self.root == other.root
+
+    def __repr__(self):
+        return "bintree(uid={},root={!r})".format(self.uid, self.root)
+
+class BinaryTreeNode(DataStructure):
+
+    def __init__(self, data, left=None, right=None, **kwargs):
+        # We don't handle storing data on the edges.  If somebody wants it, we'll add it.
```

Comment:
Personally, I would consider it a lacking feature if I couldn't visualize a tree with values on edges e.g. for decision trees. We could add it as a separate data structure later though.

Reply:
Hm... We could also toss in something ugly like an optional "edgevals" key alongside a "children" key for trees.